File: code/dart/test/metrics/metrics_integration_test.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:modular_api/modular_api.dart';
import 'package:modular_api/src/core/modular_api.dart' show apiRegistry;
import 'package:test/test.dart';

// ── Minimal UseCase for integration tests ────────────────────────────────

class _PingInput extends Input {
  _PingInput();
  factory _PingInput.fromJson(Map<String, dynamic> json) => _PingInput();
  @override
  Map<String, dynamic> toJson() => {};
  @override
  Map<String, dynamic> toSchema() => {'type': 'object', 'properties': {}};
}

class _PingOutput extends Output {
  _PingOutput();
  @override
  Map<String, dynamic> toJson() => {'pong': true};
  @override
  Map<String, dynamic> toSchema() => {
        'type': 'object',
        'properties': {
          'pong': {'type': 'boolean'},
        },
      };
  @override
  int get statusCode => 200;
}

class _PingUseCase implements UseCase<_PingInput, _PingOutput> {
  @override
  final _PingInput input;
  @override
  ModularLogger? logger;

  _PingUseCase(this.input);
  static _PingUseCase fromJson(Map<String, dynamic> json) =>
      _PingUseCase(_PingInput.fromJson(json));

  @override
  String? validate() => null;

  @override
  Future<_PingOutput> execute() async {
    return _PingOutput();
  }
}

// ── Tests ────────────────────────────────────────────────────────────────

void main() {
  group('ModularApi metrics integration', () {
    late HttpServer server;
    late String baseUrl;

    tearDown(() async {
      await server.close(force: true);
      // Clear global registry for next test
      apiRegistry.routes.clear();
    });

    Future<HttpServer> startServer({bool metricsEnabled = false}) async {
      final api = ModularApi(
        basePath: '/api',
        title: 'Test API',
        version: '1.0.0',
        metricsEnabled: metricsEnabled,
      );
      api.module('test', (m) {
        m.usecase('ping', _PingUseCase.fromJson,
            inputExample: _PingInput(), outputExample: _PingOutput());
      });
      server = await api.serve(port: 0);
      baseUrl = 'http://localhost:${server.port}';
      return server;
    }

    test('metrics disabled by default — /api/metrics returns 404', () async {
      await startServer(metricsEnabled: false);
      final response = await http.get(Uri.parse('$baseUrl/api/metrics'));
      expect(response.statusCode, equals(404));
    });

    test(
        'metrics enabled — GET /api/metrics returns 200 with prometheus content type',
        () async {
      await startServer(metricsEnabled: true);
      final response = await http.get(Uri.parse('$baseUrl/api/metrics'));
      expect(response.statusCode, equals(200));
      expect(
        response.headers['content-type'],
        contains('text/plain'),
      );
      expect(response.body, contains('process_start_time_seconds'));
    });

    test('metrics enabled — requests are instrumented', () async {
      await startServer(metricsEnabled: true);

      // Make a request to the use case endpoint
      await http.post(
        Uri.parse('$baseUrl/api/test/ping'),
        headers: {'content-type': 'application/json'},
        body: jsonEncode({}),
      );

      // Check metrics endpoint
      final metricsResponse = await http.get(Uri.parse('$baseUrl/api/metrics'));
      final body = metricsResponse.body;

      expect(body, contains('http_requests_total'));
      expect(body, contains('method="POST"'));
      expect(body, contains('status_code="200"'));
      expect(body, contains('route="/api/test/ping"'));
      expect(body, contains('http_request_duration_seconds'));
    });

    test('metrics enabled — /api/health and /api/docs are not instrumented', () async {
      await startServer(metricsEnabled: true);

      // Hit health and docs
      await http.get(Uri.parse('$baseUrl/api/health'));
      await http.get(Uri.parse('$baseUrl/api/docs'));

      // Metrics should not contain health or docs routes
      final metricsResponse = await http.get(Uri.parse('$baseUrl/api/metrics'));
      final body = metricsResponse.body;

      // Only process_start_time_seconds should be present
      // No http_requests_total lines (since only excluded routes were hit)
      expect(body, isNot(contains('route="/api/health"')));
      expect(body, isNot(contains('route="/api/docs"')));
    });

    test('metrics getter returns MetricsRegistrar when enabled', () async {
      final api = ModularApi(
        basePath: '/api',
        version: '1.0.0',
        metricsEnabled: true,
      );
      expect(api.metrics, isNotNull);
    });

    test('metrics getter returns null when disabled', () async {
      final api = ModularApi(basePath: '/api', version: '1.0.0');
      expect(api.metrics, isNull);
    });

    test('custom metrics appear in /api/metrics output', () async {
      final api = ModularApi(
        basePath: '/api',
        version: '1.0.0',
        metricsEnabled: true,
      );
      api.module('test', (m) {
        m.usecase('ping', _PingUseCase.fromJson,
            inputExample: _PingInput(), outputExample: _PingOutput());
      });

      // Register a custom metric
      final customCounter = api.metrics!.createCounter(
        name: 'custom_operations_total',
        help: 'Custom operations counter',
      );
      customCounter.labels({'type': 'test'}).inc(42);

      server = await api.serve(port: 0);
      baseUrl = 'http://localhost:${server.port}';

      final response = await http.get(Uri.parse('$baseUrl/api/metrics'));
      expect(response.body, contains('custom_operations_total'));
      expect(response.body, contains('type="test"'));
      expect(response.body, contains('42'));
    });
  });
}

File: code/ts/test/metrics/metrics_integration.test.ts
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import request from 'supertest';
import type { Server } from 'http';
import { ModularApi, Input, Output, UseCase } from '../../src';
import { apiRegistry } from '../../src/core/registry';

// ── Minimal UseCase for integration tests ────────────────────

class PingInput extends Input {
  toJson() {
    return {};
  }
  toSchema() {
    return { type: 'object', properties: {} };
  }
}

class PingOutput extends Output {
  get statusCode() {
    return 200;
  }
  toJson() {
    return { pong: true };
  }
  toSchema() {
    return {
      type: 'object',
      properties: { pong: { type: 'boolean' } },
    };
  }
}

class PingUseCase extends UseCase<PingInput, PingOutput> {
  readonly input: PingInput;

  constructor(input: PingInput) {
    super();
    this.input = input;
  }

  static fromJson(_json: Record<string, unknown>) {
    return new PingUseCase(new PingInput());
  }

  validate() {
    return null;
  }

  async execute() {
    return new PingOutput();
  }
}

// ── Tests ────────────────────────────────────────────────────

describe('ModularApi metrics integration', () => {
  let server: Server;

  afterEach(async () => {
    if (server) {
      await new Promise<void>((resolve) => server.close(() => resolve()));
    }
    apiRegistry.clear();
  });

  async function startServer(opts: { metricsEnabled?: boolean } = {}) {
    const api = new ModularApi({
      basePath: '/api',
      title: 'Test API',
      version: '1.0.0',
      metricsEnabled: opts.metricsEnabled ?? false,
    });
    api.module('test', (m) => {
      m.usecase('ping', PingUseCase.fromJson, { inputClass: PingInput, outputClass: PingOutput });
    });
    server = await api.serve({ port: 0 });
    return server;
  }

  it('metrics disabled by default — /api/metrics returns 404', async () => {
    const srv = await startServer({ metricsEnabled: false });
    const res = await request(srv).get('/api/metrics');
    expect(res.status).toBe(404);
  });

  it('metrics enabled — GET /api/metrics returns 200 with prometheus content type', async () => {
    const srv = await startServer({ metricsEnabled: true });
    const res = await request(srv).get('/api/metrics');
    expect(res.status).toBe(200);
    expect(res.headers['content-type']).toContain('text/plain');
    expect(res.text).toContain('process_start_time_seconds');
  });

  it('metrics enabled — requests are instrumented', async () => {
    const srv = await startServer({ metricsEnabled: true });

    // Make a request to the usecase endpoint
    await request(srv).post('/api/test/ping').send({}).set('Content-Type', 'application/json');

    // Check metrics
    const res = await request(srv).get('/api/metrics');
    expect(res.text).toContain('http_requests_total');
    expect(res.text).toContain('method="POST"');
    expect(res.text).toContain('status_code="200"');
    expect(res.text).toContain('route="/api/test/ping"');
    expect(res.text).toContain('http_request_duration_seconds');
  });

  it('metrics enabled — /api/health and /api/docs are not instrumented', async () => {
    const srv = await startServer({ metricsEnabled: true });

    await request(srv).get('/api/health');
    await request(srv).get('/api/docs');

    const res = await request(srv).get('/api/metrics');
    expect(res.text).not.toContain('route="/api/health"');
    expect(res.text).not.toContain('route="/api/docs"');
  });

  it('metrics getter returns MetricsRegistrar when enabled', () => {
    const api = new ModularApi({
      basePath: '/api',
      version: '1.0.0',
      metricsEnabled: true,
    });
    expect(api.metrics).toBeDefined();
  });

  it('metrics getter returns undefined when disabled', () => {
    const api = new ModularApi({ basePath: '/api', version: '1.0.0' });
    expect(api.metrics).toBeUndefined();
  });

  it('custom metrics appear in /api/metrics output', async () => {
    const api = new ModularApi({
      basePath: '/api',
      version: '1.0.0',
      metricsEnabled: true,
    });
    api.module('test', (m) => {
      m.usecase('ping', PingUseCase.fromJson, { inputClass: PingInput, outputClass: PingOutput });
    });

    // Register custom metric
    const customCounter = api.metrics!.createCounter({
      name: 'custom_operations_total',
      help: 'Custom operations counter',
      labelNames: ['type'] as const,
    });
    customCounter.inc({ type: 'test' }, 42);

    server = await api.serve({ port: 0 });

    const res = await request(server).get('/api/metrics');
    expect(res.text).toContain('custom_operations_total');
    expect(res.text).toContain('type="test"');
    expect(res.text).toContain('42');
  });
});
